server/src/auth.rs

```diff
@@ -0,0 +1,144 @@
+
+fn is_valid_user(user: &NamedUser) -> bool {
+    // this is a dummy implementation, needs some kind of config backend:
+    // i thought of an additional (read only?) store you can specify on the
+    // command-line: the store should contain, user, hashed passwords, and a
+    // link between users ans permissions/actions. but not 100% sure how to
+    // model this in rdf.
+    return user.username == "admin" && user.password == "password";
+}
+
+fn can_user_access(user: &User, required_permissions: Permission) -> bool {
+    // see is_valid_user
+    match required_permissions {
+        Permission::Query => match user {
+            User::Anonymous => false,
+            User::Named(_) => true,
+        },
+        _ => true,
+    }
+}
+
+// a similar function could be moved to oxhttp?
+fn read_user_from_request(request: &Request) -> anyhow::Result<User> {
+    match request.header(&HeaderName::AUTHORIZATION) {
+        None => Ok(User::Anonymous),
+        Some(header) => match header.to_str().unwrap().split_once(" ") {
+            None => Err(anyhow::anyhow!("Could not read Header")),
+            Some((prefix, payload)) => {
+                if prefix != "Basic" {
+                    return Err(anyhow::anyhow!("Only support Basic Auth"));
+                }
+                match decode(payload) {
+                    Err(_) => Err(anyhow::anyhow!("Could not base64 decode header")),
+                    Ok(decoded) => match String::from_utf8(decoded) {
+                        Err(_) => Err(anyhow::anyhow!("Could not utf8 decode header")),
+                        Ok(decoded_string) => match decoded_string.split_once(":") {
+                            None => Err(anyhow::anyhow!("Could not read User and Password")),
+                            Some((username, password)) => Ok(User::Named(NamedUser {
+                                username: username.into(),
+                                password: password.into(),
+                            })),
+                        },
+                    },
+                }
+            }
+        },
+    }
+}
+
+pub fn require_auth(required_permissions: Permission, request: &Request) -> Result<(), HttpError> {
+    match read_user_from_request(request) {
+        Ok(user) => {
+            if let User::Named(named_user) = &user {
+                if !is_valid_user(&named_user) {
+                    return Err((Status::UNAUTHORIZED, "forbidden".into()));
+                }
+            }
+            // only valid users and anonymous users left
+            if !can_user_access(&user, required_permissions) {
+                return match user {
+                    User::Anonymous => Err((Status::UNAUTHORIZED, "unauthorized".into())),
+                    User::Named(_) => Err((Status::FORBIDDEN, "forbidden".into())),
+                };
+            }
+            return Ok(());
+        }
+        Err(e) => Err((Status::BAD_REQUEST, format!("Bad request: {:?}", e))),
```

Comment:
Instead of a `match` you can use 
```rust
let user = read_user_from_request(request).map_err(|e|  (Status::BAD_REQUEST, format!("Bad request: {:?}", e)))?
```